contracts/mesh-provider/src/state.rs

```diff
@@ -101,6 +148,43 @@ pub struct Validator {
     pub multiplier: Decimal,
     // how active is it
     pub status: ValStatus,
+    pub rewards: ValidatorRewards
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct ValidatorRewards {
+    /// Last height we withdrew rewards and calculated them
+    pub last_height: u64,
+    /// rewards_per_token_per_block, total of rewards to be paid, per staked token per block.
+    pub total_rptpb: Uint128,
+}
+
+impl ValidatorRewards {
+    pub fn new(height: u64) -> Self {
+        ValidatorRewards {
+            last_height: height,
+            total_rptpb: Uint128::zero(),
+        }
+    }
+
+    pub fn calc_rewards(
+        &mut self,
+        rewards: Uint128,
+        total_tokens: Uint128,
+        curr_height: u64,
+    ) -> Result<(), ContractError> {
+        let height_diff = curr_height - self.last_height;
+
+        if height_diff == 0 {
+            return Err(ContractError::ValidatorRewardsCalculationWrong {});
+        }
+
+        self.total_rptpb = rewards
```

Comment:
This is not proper.

I got 5 tokens for 1000 tokens staked after 100 blocks. Thus I set `total_rptpb` to `0.00005`
Later, I receive the same payout. and would add to the other making `total_rptpb` to `0.0001`

If I got both payouts at the same time, then I would have
10 rewards for 1000 staked after 200 blocks, which would be `0.00005`.

In short, you can sum blocks_passed and sum total_received and keep accumulating there, and divide the two totals.
But you can sum the divided amounts. In other math speak:

`(a1 + a2 + ...) / (b1 + b2 + ...) != (a1/b1) + (a2/b2) + ... `

Reply:
Note: this could really benefit from unit tests.

The logic is nicely separated into on struct in state, which can be tested without any large ibc framework.

Simulate some bonds, unbonds, claims. Copying those few lines you call in stake/unstake/ibc_reward/claim_reward, but adapted to the struct only (using mock_env() or whatever)

Then you can check this math working out. Or not working out and have a test case to fix it

Reply:
Yup, totally wrong, just did the math again now.
Will look at the other contracts to understand them better.